File: solve/cpp/merge.cc
#include "common.hh"

int *b;

void merge(int *x1, int *x2, int *y1, int *y2)
{
    int *p = b, *x = x1, *y = y1;
    while (true)
    {
        if (*x <= *y)
        {
            *p++ = *x++;
            if (x > x2)
            {
                while (y <= y2) *p++ = *y++;
                break;
            }
        } 
        else 
        {
            *p++ = *y++;
            if (y > y2)
            {
                while (x <= x2) *p++ = *x++;
                break;
            }
        } 
    }
    while (y2 >= x1) *y2-- = *--p;
} 

void mergesort(int *lo, int *hi)
{
    int *p = lo + (hi - lo)/2;
    int *q = p + 1;
    if (p > lo) mergesort(lo, p);
    if (q < hi) mergesort(q, hi);
    merge(lo, p, q, hi);
}

int main(int argc, char *argv[])
{
    if (init(argc, argv) == 0)
    {
        b = new int[n];
        mergesort(a, a + n - 1);
        delete b;
        output();
        return 0;
    }
    return 1;
}

File: solve/lua/merge.lua
require "common" 

function merge(x1, x2, y1, y2) 
    local b, x, y = {  }, x1, y1
    while true do
        if a[x] <= a[y] then
            table.insert(b, a[x])
            if x == x2 then
                for i = y, y2 do
                    table.insert(b, a[i])
                end     
                break
            else
                x = x + 1
            end 
        else
            table.insert(b, a[y])
            if y == y2 then
                for i = x, x2 do
                    table.insert(b, a[i])
                end     
                break
            else
                y = y + 1
            end 
        end
    end
    local j = 1
    for i = x1, y2 do
        a[i], j = b[j], j + 1
    end
end

function mergesort(lo, hi)
    if lo < hi then
        local mid = math.floor((lo + hi)/2)
        mergesort(lo, mid)
        mergesort(mid + 1, hi)
        merge(lo, mid, mid + 1, hi)
    end 
end

skeleton(function ()
    mergesort(1, n)
end)
